xds/internal/balancer/cdsbalancer/cluster_handler_test.go

```diff
@@ -0,0 +1,390 @@
+	edsService = "EDS Service"
+	logicalDNSService = "Logical DNS Service"
+	edsService2 = "EDS Service 2"
+	logicalDNSService2 = "Logical DNS Service 2"
+	aggregateClusterService = "Aggregate Cluster Service"
+)
+
+// setupTests creates a clusterHandler with a fake xds client for control over xds client.
+func setupTests(t *testing.T) (*clusterHandler, *fakeclient.Client) {
+	xdsC := fakeclient.NewClient()
+	ch := &clusterHandler{
+		xdsClient: xdsC,
+		// This is will be how the update channel is created in cds. It will be a separate channel to the buffer.Unbounded.
+		// This channel will also be read from to test any cluster updates.
+		updateChannel: make(chan clusterHandlerUpdate, 1),
+	}
+	return ch, xdsC
+}
+
+// Simplest case: the cluster handler receives a cluster name, handler starts a watch for that cluster, xds client returns
+// that it is an EDS Cluster, not a tree, so expectation that update is written to buffer which will be read by CDS LB.
+func (s) TestUpdateRootClusterEDSSuccess(t *testing.T) {
+	ch, fakeClient := setupTests(t)
+	// When you first update the root cluster, it should hit the code path which will start a cluster node for that root.
+	// Updating the root cluster logically represents a ping from a ClientConn.
+	ch.updateRootCluster(edsService)
+
+	// Starting a cluster node involves communicating with the xdsClient, telling it to watch a cluster.
+	ctx, ctxCancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer ctxCancel()
+	gotCluster, err := fakeClient.WaitForWatchCluster(ctx)
+	if err != nil {
+		t.Fatalf("xdsClient.WatchCDS failed with error: %v", err)
+	}
+	if gotCluster != edsService {
+		t.Fatalf("xdsClient.WatchCDS called for cluster: %v, want: %v", gotCluster, edsService)
+	}
+
+	// Invoke callback with xds client with a certain clusterUpdate. Due to this cluster update filling out the whole
+	// cluster tree, as the cluster is of type EDS and not an aggregate cluster, this should trigger the ClusterHandler
+	// to write to the update buffer to update the CDS policy.
+	fakeClient.InvokeWatchClusterCallback(xdsclient.ClusterUpdate{
+		ClusterType: xdsclient.ClusterTypeEDS,
+		ServiceName: edsService,
+	}, nil)
+
+	select {
+		case chu := <-ch.updateChannel:
+			if len(chu.chu) != 1 {
+				t.Fatal("Cluster Update passed to CDS should only have one update as cluster type EDS")
+			}
+			if chu.chu[0].ClusterType != xdsclient.ClusterTypeEDS {
+				t.Fatalf("ClusterUpdate Type received: %v, ClusterUpdate wanted: %v", chu, xdsclient.ClusterTypeEDS)
+			}
+			if chu.chu[0].ServiceName != edsService {
+				t.Fatalf("ClusterUpdate ServiceName received: %v, ClusterUpdate wanted: %v", chu.chu[0].ServiceName, edsService)
+			}
+		case <-ctx.Done():
+			t.Fatal("Timed out waiting for update from updateChannel.")
+	}
+
+}
+
+// This test is the same as the prior test, expect that the cluster type is of LogicalDNS rather than EDS.
+func (s) TestUpdateRootClusterLogicalDNSSuccess(t *testing.T) {
```

Comment:
Nit: you can merge this with EDS_success, to get rid of the duplicate code.

```go
for _, tt := range []struct{
  clusterName string
  update ClusterUpdate
}{
 {edsService, {EDS, edsService}},
 {dnsService, {DNS, dnsService}},
}{
  t.Run(
    ...
    fakeClient.Invoke(tt.update)
    ...
    diff(chu.chu, []ClusterUpdate{tt.update})
  )
}
```

https://github.com/grpc/grpc-go/blob/master/xds/internal/balancer/cdsbalancer/cdsbalancer_test.go#L290

Reply:
Merged :).